File: frontend/src/lib/simulation/engine.ts
import type { Issue, District, Bill, SimResponse, Round, Vote, Speech } from "../api";
import { type Member, decideVote } from "./agents";
import { generateSpeech } from "./debate";
import { Random } from "./rng";

export interface SimulationResult {
  members: number;
  rounds: Round[];
  final_passed: boolean;
}

function sampleMembers(districts: District[], n: number, rng: Random): Member[] {
  // Weight by population
  const pops = districts.map(d => Math.max(1, d.population));
  const total = pops.reduce((a, b) => a + b, 0);
  const probs = pops.map(p => p / total);

  const members: Member[] = [];
  for (let i = 0; i < n; i++) {
    const d = rng.choice(districts, probs);
    // ideology roughly tracks district lean but with noise
    let ideology = rng.gauss(d.lean, 0.35);
    ideology = Math.max(-1.0, Math.min(1.0, ideology));

    // cosmetic party hint
    const partyHint = ideology > 0.15 ? "Blue" : (ideology < -0.15 ? "Red" : "Purple");

    members.push({
      member_id: `M-${String(i + 1).padStart(4, '0')}`,
      district: d,
      ideology,
      party_hint: partyHint
    });
  }
  return members;
}

function pickSpokespeople(members: Member[], k: number = 7): Member[] {
  const sortedM = [...members].sort((a, b) => a.ideology - b.ideology);
  if (sortedM.length <= k) return sortedM;

  // Pick representative slice: min, 25%, median, 75%, max
  const len = sortedM.length;
  const idxs = [0, Math.floor(len / 4), Math.floor(len / 2), Math.floor(3 * len / 4), len - 1];

  const chosen: Member[] = [];
  const seen = new Set<string>();

  for (const i of idxs) {
    const m = sortedM[i];
    if (!seen.has(m.member_id)) {
      chosen.push(m);
      seen.add(m.member_id);
    }
  }

  // Fill rest from middle 50% randomly
  const midStart = Math.floor(len / 4);
  const midEnd = Math.floor(3 * len / 4);
  const mid = sortedM.slice(midStart, midEnd);

  while (chosen.length < k && mid.length > 0) {
    const randIdx = Math.floor(Math.random() * mid.length);
    const m = mid[randIdx];
    if (!seen.has(m.member_id)) {
      chosen.push(m);
      seen.add(m.member_id);
    }
    // simple avoid infinite loop if all taken
    if (chosen.length < k && seen.size >= members.length) break;
  }

  return chosen.sort((a, b) => a.ideology - b.ideology);
}


// Helper for sequential async to avoid swamping local LLM
async function asyncVote(members: Member[], bill: Bill, useLlm: boolean, llmModel: string, threshold: number = 0.5): Promise<Vote> {
  let yes = 0, no = 0, abstain = 0;

  // We can process in chunks to be faster but safe
  // For now simple serial for stability
  for (const m of members) {
    const decision = await decideVote(m, bill, useLlm, llmModel);
    if (decision.vote === "yes") yes++;
    else if (decision.vote === "no") no++;
    else abstain++;
  }

  const passed = (yes / Math.max(1, yes + no)) >= threshold;
  return { yes, no, abstain, passed, threshold };
}

// Deprecated amendment logic for now
// function proposeAmendment(...) ...

export async function runSimulation(
  districts: District[],
  numMembers: number,
  rounds: number,
  bill: Bill, // CHANGED from issueVector
  useLlm: boolean,
  llmModel: string,
  seed: number | null
): Promise<SimulationResult> {
  const rng = new Random(seed);
  if (!districts || districts.length === 0) {
    throw new Error("No active districts are loaded.");
  }

  const members = sampleMembers(districts, numMembers, rng);
  // let currentIssues = { ...issueVector }; // Deleted
  let currentBill = bill; // In future we could amend the bill text
  const allRounds: Round[] = [];

  for (let r = 0; r < rounds; r++) {
    const spokes = pickSpokespeople(members, 7);
    const speeches: Speech[] = [];

    // Generate speeches
    for (const s of spokes) {
      // Calculate stance via LLM too?
      // For speed, let's reuse decision logic or separate?
      // Ideally speech generation implies a stance.
      // Let's call decideVote first.
      const decision = await decideVote(s, currentBill, useLlm, llmModel);

      let stance: "support" | "oppose" | "amend" = "amend";
      if (decision.vote === "yes") stance = "support";
      if (decision.vote === "no") stance = "oppose";

      // We pass a dummy issue vector or update generateSpeech to take Bill
      // TODO: Update generateSpeech signature next. 
      // For now passing empty dict to satisfy type if not changed yet, but I should update debate.ts too.
      // Actually I will update debate.ts in next step.
      // I'll assume generateSpeech signature is updated to take Bill.
      const txt = await generateSpeech(s, stance, currentBill, useLlm, llmModel);

      speeches.push({
        member_id: s.member_id,
        stance,
        text: txt,
        rationale: {} // No mathematical rationale anymore
      });
    }

    const voteResult = await asyncVote(members, currentBill, useLlm, llmModel, 0.5);
    allRounds.push({
      round_index: r,
      speeches,
      vote: voteResult
    });

    // If failed and rounds remain, we technically stop or amend.
    // Without amendment logic, we just stop if passed or fail?
    // User requested realistic logic.
    // If it fails, it fails.
    if (voteResult.passed) break;

    // If not passed, we loop? But result will be same without amendment.
    // So let's break to avoid duplicate rounds.
    if (!voteResult.passed) break;
  }

  const finalPassed = allRounds.length ? allRounds[allRounds.length - 1].vote.passed : false;
  return {
    members: numMembers,
    rounds: allRounds,
    final_passed: finalPassed
  };
}

File: backend/app/sim/engine.py
from __future__ import annotations
import random, math
from typing import Dict, List, Tuple

from .agent import District, Member, Issue
from .debate import generate_speech

def sample_members(districts: List[District], n: int, rng: random.Random) -> List[Member]:
    # Weight by population to approximate representational density.
    pops = [max(1, d.population) for d in districts]
    total = float(sum(pops))
    probs = [p/total for p in pops]

    members: List[Member] = []
    for i in range(n):
        d = rng.choices(districts, weights=probs, k=1)[0]
        # ideology roughly tracks district lean but with noise
        ideology = max(-1.0, min(1.0, rng.gauss(mu=d.lean, sigma=0.35)))
        party_hint = "Blue" if ideology > 0.15 else "Red" if ideology < -0.15 else "Purple"
        members.append(Member(member_id=f"M-{i+1:04d}", district=d, ideology=ideology, party_hint=party_hint))
    return members

def pick_spokespeople(members: List[Member], k: int = 7) -> List[Member]:
    # Choose a small set for debate to keep compute reasonable.
    sorted_m = sorted(members, key=lambda m: m.ideology)
    if len(sorted_m) <= k:
        return sorted_m
    idxs = [0, len(sorted_m)//4, len(sorted_m)//2, 3*len(sorted_m)//4, len(sorted_m)-1]
    # Keep uniqueness by member_id (Member is not hashable).
    chosen: List[Member] = []
    seen = set()
    for i in idxs:
        m = sorted_m[i]
        if m.member_id not in seen:
            chosen.append(m)
            seen.add(m.member_id)

    mid = sorted_m[len(sorted_m)//4 : 3*len(sorted_m)//4]
    while len(chosen) < k and mid:
        m = random.choice(mid)
        if m.member_id not in seen:
            chosen.append(m)
            seen.add(m.member_id)
    return chosen

def vote(members: List[Member], issue_vector: Dict[Issue, float], threshold: float = 0.5) -> Tuple[int,int,int,bool]:
    yes=no=abstain=0
    for m in members:
        u = m.utility(issue_vector)
        # abstain if near-indifferent
        if abs(u) < 0.03:
            abstain += 1
        elif u > 0:
            yes += 1
        else:
            no += 1
    passed = (yes / max(1, (yes+no))) >= threshold
    return yes, no, abstain, passed

def propose_amendment(issue_vector: Dict[Issue, float], members: List[Member]) -> Dict[Issue, float]:
    # Simple "median pull": move each issue a bit toward the median member's ideology sign.
    sorted_m = sorted(members, key=lambda m: m.ideology)
    median = sorted_m[len(sorted_m)//2].ideology if sorted_m else 0.0
    amended = dict(issue_vector)
    for k, v in issue_vector.items():
        # pull v toward 0 if median opposes the direction
        step = 0.12
        if (v > 0 and median < 0) or (v < 0 and median > 0):
            amended[k] = v * (1 - step)
        else:
            amended[k] = max(-1.0, min(1.0, v + step * (0.0 if abs(v) < 0.2 else math.copysign(0.08, v))))
    return amended

async def run_simulation(
    districts: List[District],
    num_members: int,
    rounds: int,
    issue_vector: Dict[Issue, float],
    use_llm: bool,
    llm_model: str,
    seed: int | None
):
    rng = random.Random(seed)
    if not districts:
        raise ValueError("No active districts are loaded.")
    members = sample_members(districts, num_members, rng)

    current = dict(issue_vector)
    all_rounds = []

    for r in range(rounds):
        spokes = pick_spokespeople(members, k=7)
        speeches = []
        # Determine stance by utility sign
        for s in spokes:
            u = s.utility(current)
            stance = "support" if u > 0.05 else "oppose" if u < -0.05 else "amend"
            txt = await generate_speech(s, stance, current, use_llm=use_llm, llm_model=llm_model)
            speeches.append({
                "member_id": s.member_id,
                "stance": stance,
                "text": txt,
                "rationale": s.utility_by_issue(current),
            })

        yes, no, abstain, passed = vote(members, current, threshold=0.5)
        all_rounds.append({
            "round_index": r,
            "speeches": speeches,
            "vote": {
                "yes": yes, "no": no, "abstain": abstain,
                "passed": passed, "threshold": 0.5
            }
        })

        # If it didn't pass and more rounds remain, amend toward median.
        if (not passed) and (r < rounds - 1):
            current = propose_amendment(current, members)

    final_passed = all_rounds[-1]["vote"]["passed"] if all_rounds else False
    return {
        "members": num_members,
        "rounds": all_rounds,
        "final_passed": final_passed,
    }
